table.cc

```diff
@@ -2234,26 +2233,18 @@ future<row_locker::lock_holder> table::do_push_view_replica_updates(const schema
     // We'll return this lock to the caller, which will release it after
     // writing the base-table update.
     future<row_locker::lock_holder> lockf = local_base_lock(base, m.decorated_key(), slice.default_row_ranges(), timeout);
-    return utils::get_local_injector().inject("table_push_view_replica_updates_timeout", timeout).then([lockf = std::move(lockf), timeout] () mutable {
-        return std::move(lockf);
-    }).then([m = std::move(m), slice = std::move(slice), views = std::move(views), base, this, timeout, now, source = std::move(source), &sem, tr_state = std::move(tr_state), &io_priority] (row_locker::lock_holder lock) mutable {
-      return do_with(
-        dht::partition_range::make_singular(m.decorated_key()),
-        std::move(slice),
-        std::move(m),
-        [base, views = std::move(views), lock = std::move(lock), this, timeout, now, source = std::move(source), &sem, &io_priority, tr_state = std::move(tr_state)] (auto& pk, auto& slice, auto& m) mutable {
-            auto permit = sem.make_permit(base.get(), "push-view-updates-2");
-            auto reader = source.make_reader(base, permit, pk, slice, io_priority, tr_state, streamed_mutation::forwarding::no, mutation_reader::forwarding::no);
-            return this->generate_and_propagate_view_updates(base, std::move(permit), std::move(views), std::move(m), std::move(reader), tr_state, now)
-                    .then([base, tr_state = std::move(tr_state), lock = std::move(lock)] () mutable {
-                tracing::trace(tr_state, "View updates for {}.{} were generated and propagated", base->ks_name(), base->cf_name());
-                // return the local partition/row lock we have taken so it
-                // remains locked until the caller is done modifying this
-                // partition/row and destroys the lock object.
-                return std::move(lock);
-            });
-      });
-    });
+    co_await utils::get_local_injector().inject("table_push_view_replica_updates_timeout", timeout);
+    auto lock = co_await std::move(lockf);
+    auto pk = dht::partition_range::make_singular(m.decorated_key());
+    auto permit = sem.make_permit(base.get(), "push-view-updates-2");
+    auto reader = source.make_reader(base, permit, pk, slice, io_priority, tr_state, streamed_mutation::forwarding::no, mutation_reader::forwarding::no);
+    co_await this->generate_and_propagate_view_updates(base, std::move(permit), std::move(views), std::move(m), std::move(reader), tr_state, now);
+    tracing::trace(tr_state, "View updates for {}.{} were generated and propagated", base->ks_name(), base->cf_name());
+    // return the local partition/row lock we have taken so it
+    // remains locked until the caller is done modifying this
+    // partition/row and destroys the lock object.
+    co_return std::move(lock);
+
 }
```

Comment:
So beautiful, I'm crying.

Reply:
Yes, it's much nicer than the old continuations, BUT, if you're fixing here a real bug that will need backporting, we'll have big problems backporting changes based on this patch to versions that can't use coroutines :-(

Reply:
I think it's still worth it (speaking as a person most likely responsible for these painful backports). Backporting would require being extra careful, but on the other hand this cleaner code makes future bugs less likely to appear.